### PR TITLE
Nat/python deployment

### LINKED ISSUE #642: Smart defaults for Python
I'll work with the Python team to better understand the heuristics and defaults.

### PR DESCRIPTION
**Pull request by nturinski**
Fixes #642 in combination with https://github.com/Microsoft/vscode-azuretools/pull/352

For zipignore, I was wondering if I should just import [this](https://github.com/github/gitignore/blob/master/Python.gitignore) `.gitignore` file and parse it. I assume anything that isn't saved in a git repo shouldn't be deployed either?

The `.deployment` file will trigger a build, Kudu detects that it is Python, and deployment works with zipdeploy.  Local git will work as is and does not require a `.deployment` file.

The deployment will break without the scmDoBuildDeploy flag, so I just enable it rather than prompting the user in this case.

### CODE REVIEW

File: src/constants.ts [hunk 1]
```diff
@@ -15,13 +15,16 @@ export enum runtimes {
...
+    javase = 'java|8-jre8',
+    python = 'python'
 }
 
 export function getIgnoredFoldersForDeployment(runtime: string): string[] {
     switch (runtime) {
         case runtimes.node:
             return ['node_modules{,/**}'];
+        case runtimes.python:
+            return ['.env', '.Python', 'env/{,/**}', 'build{,/**}', 'dist{,/**}'];
```
Comment (EricJizbaMSFT) [inline src/constants.ts]: Why do some folders have `{,/**}` and some don't?

File: src/commands/deploy.ts [hunk 2]
```diff
@@ -115,10 +115,16 @@ export async function deploy(context: IActionContext, confirmDeployment: boolean
...
-                // tslint:disable-next-line:no-unsafe-any
-                await node.enableScmDoBuildDuringDeploy(fsPath, constants.runtimes[siteConfig.linuxFxVersion.substring(0, siteConfig.linuxFxVersion.indexOf('|'))], context.properties);
+            //check if node is being zipdeployed and that there is no .deployment file
+            if (siteConfig.linuxFxVersion && siteConfig.scmType === 'None' && !(await pathExists(path.join(fsPath, constants.deploymentFileName)))) {
+                if (siteConfig.linuxFxVersion.startsWith(constants.runtimes.node)) {
+                    // if it is node, prompt the user (as we can break them)
+                    await node.promptScmDoBuildDeploy(fsPath, constants.runtimes[constants.runtimes.node], context.properties);
+                } else if (siteConfig.linuxFxVersion.startsWith(constants.runtimes.python)) {
+                    // if it is a python app, enable it for them as it does not work without build during deploy
+                    await node.enableScmDoBuildDuringDeploy(fsPath, constants.runtimes.python);
```
Comment (EricJizbaMSFT) [inline src/commands/deploy.ts]: > it does not work without build during deploy

Can you point to docs for this? And will this always be the case? I don't like this for two reasons:
1. It happens on every deploy and users have no way to override the behavior if they don't want it
1. I think it's confusing when tooling changes a user's project without telling them

File: src/constants.ts (hunk 1 above)
Comment (EricJizbaMSFT) [inline src/constants.ts]: How did you pick these virtual environment names? Couldn't users name it almost anything?

Comment (EricJizbaMSFT) [reply src/constants.ts]: `.env` is one of the most common names for python virtual environments. It's definitely a folder

Comment (nturinski) [reply src/constants.ts]: I thought `.env` was a file describing environment variables https://www.ibm.com/support/knowledgecenter/en/ssw_aix_72/com.ibm.aix.osdevice/env_file.htm where env is the folder for virtual environments

File: src/commands/deploy.ts (hunk 2 above)
Comment (nturinski) [reply src/commands/deploy.ts]: Okay so, I tried creating the venv and then pip install the requirements and deployed that and it worked.  I'll handle it the same way we handle node.js.

File: src/constants.ts [hunk 3]
```diff
@@ -15,13 +15,16 @@ export enum runtimes {
...
-    javase = 'java|8-jre8'
+    javase = 'java|8-jre8',
+    python = 'python'
 }
 
 export function getIgnoredFoldersForDeployment(runtime: string): string[] {
     switch (runtime) {
         case runtimes.node:
             return ['node_modules{,/**}'];
+        case runtimes.python:
```
Comment (EricJizbaMSFT) [inline src/constants.ts]: I still think users name a virtual environment whatever they want. Ideally we're smarter about this. For example, you can "detect" a venv by check the path for an activate script. Something similar to https://github.com/Microsoft/vscode-azurefunctions/blob/master/src/commands/createNewProject/PythonProjectCreator.ts#L200

That being said, I'm fine if you hard-code the most common "envs" here and file an issue to be smarter later

File: src/constants.ts [hunk 4]
```diff
@@ -15,13 +15,16 @@ export enum runtimes {
...
+            return ['.env', 'env/{,/**}', 'build{,/**}', 'dist{,/**}'];
```
[the remark this answers is not in the document]
Comment (nturinski) [reply src/constants.ts]: Yeah, that's a good point.  I can't get a definitive answer whether or not to ignore `.env` files on a deploy.  Furthermore, I think some developers use `dotenv` for production code, so their `.env` file might be required anyway.

[the remark this answers is not in the document]
Comment (nturinski) [reply src/constants.ts]: Also did some testing to confirm, but you do need the `/{,/**}` for it to ignore the directory.

File: src/constants.ts (hunk 3 above)
Comment (nturinski) [reply src/constants.ts]: Yeah, fair enough.  I'll add the environments that the .gitignore lists for now and make a smarter solution in a separate PR.

File: src/constants.ts [hunk 5]
```diff
@@ -15,13 +15,16 @@ export enum runtimes {
     dotnetcore = 'dotnetcore',
     ruby = 'ruby',
     tomcat = 'tomcat',
-    javase = 'java|8-jre8'
+    javase = 'java|8-jre8',
+    python = 'python'
 }
 
 export function getIgnoredFoldersForDeployment(runtime: string): string[] {
```
[the remark this answers is not in the document]
Comment (nturinski) [reply src/constants.ts]: 😬 Ugh, what was I _thinking?!_

**Comment by nturinski** [comment]
I added a new check to see if the folder we are going to add to the `zipIgnorePattern` exists within the project.  A potential downfall could be if a user adds `node_modules` after the first deployment, we wouldn't catch it, but we also wouldn't add a bunch of unnecessary settings that could be confusing.

I'm thinking it might not be worth adding but wanted your opinion about it.

File: src/explorer/SiteTreeItem.ts [hunk 6]
```diff
@@ -166,6 +157,28 @@ export abstract class SiteTreeItem extends AzureParentTreeItem<ISiteTreeRoot> {
...
+        const zipIgnoreFolders: string[] = this.getIgnoredFoldersForDeployment(runtime);
+        let oldSettings: string[] | string | undefined = workspace.getConfiguration(constants.extensionPrefix, Uri.file(fsPath)).get(constants.configurationSettings.zipIgnorePattern);
+        if (!oldSettings) {
+            oldSettings = [];
+        } else if (typeof oldSettings === "string") {
+            oldSettings = [oldSettings];
+            // settings have to be an array to concat the proper zipIgnoreFolders
+        }
+        const newSettings: string[] = oldSettings;
+        for (const folder of zipIgnoreFolders) {
```
Comment (EricJizbaMSFT) [inline src/explorer/SiteTreeItem.ts]: Yeah I don't like the check for 'pathExists'. There's no harm in ignoring a folder that doesn't exist

File: src/explorer/SiteTreeItem.ts [hunk 7]
```diff
@@ -182,4 +195,15 @@ export abstract class SiteTreeItem extends AzureParentTreeItem<ISiteTreeRoot> {
             telemetryProperties.promptToSaveDeployConfigs = 'Skip for now';
         }
     }
+
+    private getIgnoredFoldersForDeployment(runtime: string): string[] {
+        switch (runtime) {
+            case runtimes.node:
+                return ['node_modules{,/**}'];
+            case runtimes.python:
+                return ['env/{,/**}', 'venv{,/**}', 'lib{,/**}', 'lib64{,/**}'];
```
Comment (EricJizbaMSFT) [inline src/explorer/SiteTreeItem.ts]: You keep changing this list - how are you picking these??

Comment (nturinski) [reply src/explorer/SiteTreeItem.ts]: Just looking at a .gitignore file on GitHub and picking ones that seem relevant (like pip installs into a Lib folder, as far as I can tell) https://github.com/github/gitignore/blob/master/Python.gitignore

Comment (EricJizbaMSFT) [reply src/explorer/SiteTreeItem.ts]: I would prefer that all envs listed in the gitignore were included here before checking in (on the off chance that we don't implement the "auto-detect" before releasing)
aka all of these:
```
# Environments
.env
.venv
env/
venv/
ENV/
env.bak/
venv.bak/
```

File: src/explorer/SiteTreeItem.ts [hunk 8]
```diff
@@ -182,4 +192,15 @@ export abstract class SiteTreeItem extends AzureParentTreeItem<ISiteTreeRoot> {
...
```
Comment (EricJizbaMSFT) [inline src/explorer/SiteTreeItem.ts]: 'env/{,/**}' has an extra slash

File: src/explorer/SiteTreeItem.ts [hunk 9]
```diff
@@ -182,4 +192,15 @@ export abstract class SiteTreeItem extends AzureParentTreeItem<ISiteTreeRoot> {
...
+            default:
```
Comment (EricJizbaMSFT) [inline src/explorer/SiteTreeItem.ts]: Please file an issue and/or send an email to the python team about the non-virtual environment values as well. You picked two values (lib and lib64) out of a pretty long list from the "# Distribution / packaging" section and it still seems arbitrary. I also think we might want to include the values from the "# Byte-compiled / optimized / DLL files" section

**Comment by nturinski** [comment]
When you say the list of changed files will change before release, are you referring to the non-virtual environment values or the virtual environment detection logic or both?

**Comment by EricJizbaMSFT** [comment]
Both. For virtual environment values I don't think auto-detection is required for release, but the list of hard-coded values should at least be longer than two values in that case

**Pull request merged**